<div class="panel-body">
  <div class="text-explanation">
    This page gives access to the data generated in <a href="http://elifesciences.org/content/4/e11282">Zanini et al. eLIFE, <i>e11282</i></a>. <br>The data are available in two forms:
  </div>
  <div class="col-xs-6">
    <h4 class="subheader"><span class="glyphicon glyphicon-circle-arrow-down"></span> Download</h4>
    <hr>
    There are three ways to query the data set:
    <ol>
      <li><b>Clinical and summary data:</b> Tables containing clinical information (viral load, CD4+ counts, HLA types, estimates of time of infection), the patient reference sequence for mapping, and the single nucleotide variant (SNP) frequency tables.</li>
      <li><b>Haplotypes:</b> Extended genetic variants, including minor ones, covering standard or custom genomic regions.</li>
      <li><b>Sequencing reads:</b> The filtered sequencing reads, mapped against the single fragments of the patient-specific reference. Download the genbank (GB) version of the reference to find PCR fragment annotations.</li>
      <li><b>Counts of pairs of SNPs:</b> The counts of pairs of single nucleotide polymorphisms, for each sample and amplicon. These data are useful to investigate linkage.</li>
    </ol>
  </div>
  <div class="col-xs-6">
    <h4 class="subheader"><span class="glyphicon glyphicon-random"></span> RESTful API</h4>
    <hr>
    The API is used for online analysis or embedding into another website. See <a href="#api">below</a> for details.
  </div>
</div>
